platform/cxxsupport/mstd_algorithm

```diff
@@ -0,0 +1,385 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MSTD_ALGORITHM_
+#define MSTD_ALGORITHM_
+
+/* <mstd_algorithm>
+ *
+ * - provides <algorithm>
+ * - For ARM C 5, standard C++11/14 features:
+ *   - std::min, std::max for std::initializer_list
+ *   - std::all_of, std::any_of, std::none_of
+ *   - std::find_if_not
+ *   - std::equal (2-range forms)
+ *   - std::copy_n, std::move, std::move_backward
+ *   - mstd::min, mstd::max constexpr replacements
+ */
+
+#include <algorithm>
+
+#ifdef __CC_ARM
+#include <mstd_utility>
+#endif
+
+namespace mstd {
+#ifdef __CC_ARM
+// Really want basic min/max to be constexpr as per C++14
+template <typename T>
+constexpr const T &max(const T &a, const T &b)
+{
+    return a < b ? b : a;
+}
+
+template <typename T, class Compare>
+constexpr const T &max(const T &a, const T &b, Compare comp)
+{
+    return comp(a, b) ? b : a;
+}
+
+template <typename T>
+constexpr const T &min(const T &a, const T &b)
+{
+    return b < a ? b : a;
+}
+
+template <typename T>
+constexpr const T &min(const T &a, const T &b, Compare comp)
+{
+    return comp(b, a) ? b : a;
+}
+
+// Maybe sort out C++14 constexpr for these later - these are C++11 at least
+template <typename T>
+T max(initializer_list<T> il)
+{
+    return *std::max_element(begin(il), end(il));
```

Comment:
Last thing we want to do in embedded space is be recursive. And surely we can assume `max_element` is already efficient?

Recursion is certainly a thing for "make it constexpr for C++11", but if it ever did fall back to run-time, we'd be doomed. So generally I want to steer clear of constexpr if it means potential run-time recursion. We don't get to have 2 implementations :(

On the whole, I'm not /that/ interested in tuning ARM C 5 for performance, it's more just trying to get it off the ground and not break the build.

But there are places where constexpr is useful - we're currently missing `aligned_union`, which might be wanted as a basis for `variant`, and its implementation might want to do a constexpr `max` on a pack of `sizeof` via an initializer list. So might end up changing that one. I guess we can assume no-one is passing around non-constant initializer-lists, right? (They're not supposed to be able to construct them themselves).

Reply:
Thanks for the explanations. I should have clarified my question. I wanted to know if the compiler is able to do that at compile time when the list of Ts is known. It was curiosity more than anything else 😄 .

Reply:
Ah, right. I believe it won't do it at compile time, no. Because the existing `max_element` is a straightforward loop, and that's beyond the compile-time capabilities of the compiler.

But a recursive constexpr implementation would be possible, but would be less efficient if it was ever computed for non-constant arguments.

C++11's constexpr restrictions forces you to write weird algorithms to achieve constexprness. You don't generally want to do it, but for the specific case of initializer_lists which are normally (but not necessarily) constant, you might want to do it.